crowbar_framework/app/controllers/api/upgrade_controller.rb

```diff
@@ -64,6 +64,24 @@ def services
     end
   end
 
+  api :POST, "/api/upgrade/controllers", "Orchestrate the upgrade of controller nodes"
+  api_version "2.0"
+  def controllers
+    # FIXME: we need to go cluster by cluster...?
+    # First node to upgade is DRBD slave
+    # FIXME: this actually does all drbd slaves in all clusters, do we want that?
+    NodeObject.find("drbd_rsc_postgresql_master:true").each do |node|
+      node_api = Api::Node.new
+      next if node_api.already_upgraded? node
+      node_api.pre_upgrade node
```

Comment:
Not necessarily. I think the whole method for installing controllers should be asynchronous